rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+    _ => {}
+  }
+}
+```
+
+this would produce, when compiled, the following ABI specification file:
+
+```json
+
+{
+  "programType": "script",
+  "specVersion": "1",
+  "encodingVersion": "1",
+  "concreteTypes": [
+    {
+      "type": "()",
+      "concreteTypeId": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d"
+    },
+    {
+      "type": "enum MyError",
+      "concreteTypeId": "44781f4b1eb667f225275b0a1c877dd4b9a8ab01f3cd01f8ed84f95c6cd2f363",
+      "metadataTypeId": 1
+    },
+    {
+      "type": "u8",
+      "concreteTypeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
+    }
+  ],
+  "metadataTypes": [
+    {
+      "type": "(_, _)",
+      "metadataTypeId": 0,
+      "components": [
+        {
+          "name": "__tuple_element",
+          "typeId": 2
+        },
+        {
+          "name": "__tuple_element",
+          "typeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
+        }
+      ]
+    },
+    {
+      "type": "enum MyError",
+      "metadataTypeId": 1,
+      "components": [
+        {
+          "name": "A",
+          "errorMessage": "error A",
+          "typeId": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d"
+        },
+        {
+          "name": "B",
+          "errorMessage": "error B",
+          "typeId": 0
+        }
+      ]
+    },
+    {
+      "type": "u64",
+      "metadataTypeId": 2
+    }
+  ],
+  "functions": [
+    {
+      "inputs": [
+        {
+          "name": "a",
+          "concreteTypeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
+        }
+      ],
+      "name": "main",
+      "output": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d",
+      "attributes": null
+    }
+  ],
+  "loggedTypes": [
+    {
+      "logId": "4933727799282657266",
+      "concreteTypeId": "44781f4b1eb667f225275b0a1c877dd4b9a8ab01f3cd01f8ed84f95c6cd2f363"
+    }
+  ],
+  "messagesTypes": [],
+  "configurables": [],
+  "errorCodes": {
+    "18446744069414584320": {
+        "pos": {
+          "file": "main.sw",
+          "line": 13,
+          "col": 10
+        },
+        "logId": null,
+        "msg": "a str panic"
+    },
+    "18446744069414584321": {
+        "pos": {
+          "file": "main.sw",
+          "line": 14,
+          "col": 10
+        },
+        "logId": null,
+        "msg": "another str panic"
+    },
+    "18446744069414584322": {
+        "pos": {
+          "file": "main.sw",
+          "line": 15,
+          "col": 10
+        },
+        "logId": "4933727799282657266",
+        "msg": null,
+    },
+    "18446744069414584323": {
+        "pos": {
+          "file": "main.sw",
+          "line": 16,
+          "col": 10
+        },
+        "logId": "4933727799282657266",
+        "msg": null,
+    }
+    "18446744069414584324": {
+        "pos": {
+          "file": "main.sw",
+          "line": 17,
+          "col": 10
+        },
```

Comment:
It would be beneficial to have a `"fn": "function-name"` field in the `pos` object as it'd allow the SDKs to inform the user of the failure modes of the method that's being invoked, be it through some sort of `Result<T, Err>` type or adding comments in the SDK-generated code.

Reply:
Can you expand on what you mean here? The function the `panic` is within? The contract method that invoked the `panic`?

Given that functions don't exist at runtime and that a given `panic` invocation can be invoked from multiple places, this is ambiguous and may not be decidable information.

I understand the need for some kind of stack trace down the line but I'm not sure how achievable that even is.

Reply:
> Can you expand on what you mean here? The function the panic is within? The contract method that invoked the panic?

I mean the contract method that invoked the `panic` as that's the thing that users interact with via the the SDK. Perhaps that information could even be put as an array in the function part of the abi:
```
  "functions": [
    {
      "inputs": [
        {
          "name": "a",
          "concreteTypeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
        }
      ],
      "name": "main",
      "output": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d",
      "attributes": null,
      "errorCodes": [ "18446744069414584321", "18446744069414584321", 
                      "18446744069414584322", "18446744069414584323", 
                      "18446744069414584324", "18446744069414584325" ]
    }
  ],
```

Reply:
> It's not clear to me what an hypothetical fn field would contain here and how helpful it would be to provide that information.

Just the contract function's name so that the SDK can inform the user that the function can panic with the error. This seemed like an opportunity to provide more info to the user, but if it can't be done - it can't be done.